src/runtime_src/core/common/xdp/profile.cpp

```diff
@@ -441,289 +471,127 @@ update_device(void* handle, bool hw_context_flow)
-    }
-    try {
-      xrt_core::xdp::aie::status::update_device(handle, hw_context_flow);
-    } catch (const std::exception &e) {
-      std::stringstream msg;
-      msg << "Failed to setup for AIE Status. Caught exception " << e.what();
-      xrt_core::message::send(xrt_core::message::severity_level::debug, "XRT", msg.str());
-    }
-  }
-
-  if (xrt_core::config::get_aie_profile()) {
-    if (xrt_core::config::get_xdp_mode() == "xdna") {
-      xrt_core::message::send(xrt_core::message::severity_level::debug, "XRT",
-        "xdp_mode config is set to XDNA. Hence, profiling will be available only for XDNA device.");
-
-      try {
-        xrt_core::xdp::aie::profile::load_xdna();
-      }
-      catch (const std::exception &e) {
-        std::stringstream msg;
-        msg << "Failed to load AIE Profile library for XDNA mode. Caught exception " << e.what();
-        xrt_core::message::send(xrt_core::message::severity_level::debug, "XRT", msg.str());
-      }
-
-      try {
-        xrt_core::xdp::aie::profile::update_device(handle, hw_context_flow);
-      }
-      catch (const std::exception &e) {
-        std::stringstream msg;
-        msg << "Failed to setup for AIE Profile XDNA. Caught exception " << e.what();
-        xrt_core::message::send(xrt_core::message::severity_level::debug, "XRT", msg.str());
-      }
-    }
-    else {
-      xrt_core::message::send(xrt_core::message::severity_level::debug, "XRT",
-        "xdp_mode config is set to ZOCL. Hence, profiling will be available only for ZOCL device.");
-      try {
-        xrt_core::xdp::aie::profile::load();
-      }
-      catch (const std::exception &e) {
-        std::stringstream msg;
-        msg << "Failed to load AIE Profile library for ZOCL mode. Caught exception " << e.what();
-        xrt_core::message::send(xrt_core::message::severity_level::debug, "XRT", msg.str());
-      }
-
-      try {
-        xrt_core::xdp::aie::profile::update_device(handle, hw_context_flow);
-      }
-      catch (const std::exception &e) {
-        std::stringstream msg;
-        msg << "Failed to setup for AIE Profile ZOCL. Caught exception " << e.what();
-        xrt_core::message::send(xrt_core::message::severity_level::debug, "XRT", msg.str());
-      }
-    }
-  }
+  load_once_and_update(xrt_core::config::get_ml_timeline,
+		       xrt_core::xdp::ml_timeline::load,
+		       xrt_core::xdp::ml_timeline::update_device,
+		       "Failed to load ML Timeline library. Caught exception ",
+		       "Failed to setup for ML Timeline. Caught exception ",
+		       handle);
+
+  load_once_and_update(xrt_core::config::get_aie_halt,
+		       xrt_core::xdp::aie::halt::load,
+		       xrt_core::xdp::aie::halt::update_device,
+		       "Failed to load AIE Halt library. Caught exception ",
+		       "Failed to setup for AIE Halt. Caught exception ",
+		       handle);
+
+  load_once_and_update(xrt_core::config::get_aie_trace,
```

Comment:
For VE2, loading of AIE Trace Plugin library is also dependent on "xdp_mode". Please check the "xdp_mode" config and load the appropriate AIE Trace library here.